lib/ContractInteract/TokenHolder.js

```diff
@@ -0,0 +1,185 @@
+    return this.contract.methods.authorizeSession(sessionKey, spendingLimit, expirationHeight).encodeABI();
+  }
+
+  /**
+   * Revokes session for the specified session key.
+   *
+   * @param {String} sessionKey Session key to revoke.
+   *
+   * @returns {String} Executable data to revoke a session.
+   */
+  getRevokeSessionExecutableData(sessionKey) {
+    return this.contract.methods.revokeSession(sessionKey).encodeABI();
+  }
+
+  /**
+   * Logout all authorized sessions.
+   *
+   * @returns {String} Executable data to logout sessions.
+   */
+  getLogoutExecutableData() {
+    return this.contract.methods.logout().encodeABI();
+  }
+
+  /**
+   * It is used to get call prefix of executeRule method in TokenHolder contract.
+   *
+   * @returns {String} Encoded signature of executeRule method.
+   */
+  getTokenHolderExecuteRuleCallPrefix() {
+    const executeRuleHash = this.auxiliaryWeb3.utils.soliditySha3(
+      'executeRule(address,bytes,uint256,uint8,bytes32,bytes32)'
+    );
+    const executeRuleCallPrefix = executeRuleHash.substring(0, 10);
+    return executeRuleCallPrefix;
+  }
+
+  /**
+   * It is used to execute executable data signed by a session key.
+   *
+   * @param {String} to The target contract address the transaction will be executed upon.
+   * @param {String} data The payload of a function to be executed in the target contract.
+   * @param {String} nonce The nonce of an session key that was used to sign the transaction.
+   * @param {String} r `r` part of the signature.
+   * @param {String} s `s` part of the signature.
+   * @param {String} v `v` part of the signature.
+   * @param {Object} txOptions Tx options.
+   *
+   * @returns {Promise<Object>} Promise that resolves to transaction receipt.
+   */
+  async executeRule(to, data, nonce, r, s, v, txOptions) {
+    if (!txOptions) {
+      const err = new TypeError('Invalid transaction options.');
+      return Promise.reject(err);
+    }
+    if (!Web3.utils.isAddress(txOptions.from)) {
+      const err = new TypeError(`Invalid from address: ${txOptions.from}.`);
+      return Promise.reject(err);
+    }
+    const txObject = await this.executeRuleRawTx(to, data, nonce, r, s, v);
+    return Utils.sendTransaction(txObject, txOptions);
+  }
+
+  /**
+   * Private method which is used to execute executable data signed by a session key.
+   *
+   * @param {String} to The target contract address the transaction will be executed upon.
+   * @param {String} data The payload of a function to be executed in the target contract.
+   * @param {String} nonce The nonce of an session key that was used to sign the transaction.
+   * @param {String} r `r` part of the signature.
+   * @param {String} s `s` part of the signature.
+   * @param {String} v `v` part of the signature.
+   *
+   * @returns {Object} Raw transaction object.
+   */
+  executeRuleRawTx(to, data, nonce, r, s, v) {
+    if (!Web3.utils.isAddress(to)) {
```

Comment:
Also check whether `data`, `nonce`, `r`, `s`, `v` are non-empty.

Reply:
Added validation for data, nonce.
r, s, v needs wider discussion as the changes will be applied to recovery interacts also.
I will create ticket to address this later.